server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -45,11 +45,23 @@
         Property.NodeScope
     );
 
+    // For historical reason, default value is false, skip check host name check if host address exist
+    public static final Setting<Boolean> CLUSTER_ROUTING_ALLOCATION_SAME_HOST_ALWAYS_CHECK_HOSTNAME = Setting.boolSetting(
+        "cluster.routing.allocation.same_shard.host.always_check_host_name",
```

Comment:
I think it would make more sense to have this as a separate mode for the same-shard check:

```suggestion
        "cluster.routing.allocation.same_shard.host_name",
```

The intention being that users would either set `cluster.routing.allocation.same_shard.host` if they want Elasticsearch to distinguish by address, and `cluster.routing.allocation.same_shard.host_name` if they want it to distinguish hosts by name.

We also need docs for this new setting.

Reply:
I rename the setting and add docs for it. Does it meet your expectation?